src/main/java/org/joshsim/engine/external/cog/CogReader.java

```diff
@@ -0,0 +1,144 @@
+package org.joshsim.engine.external.cog;
+
+import java.io.File;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.List;
+import org.geotools.api.geometry.Position;
+import org.geotools.api.parameter.ParameterValueGroup;
+import org.geotools.api.referencing.crs.CoordinateReferenceSystem;
+import org.geotools.coverage.grid.GridCoverage2D;
+import org.geotools.coverage.processing.CoverageProcessor;
+import org.geotools.coverage.processing.Operations;
+import org.geotools.gce.geotiff.GeoTiffReader;
+import org.geotools.geometry.Position2D;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.geotools.referencing.CRS;
+import org.joshsim.engine.geometry.EngineGeometry;
+
+/**
+ * Reader for Cloud Optimized GeoTIFF (COG) files that extracts data
+ * based on specified geometries.
+ */
+public class CogReader {
+  static final CoverageProcessor processor = CoverageProcessor.getInstance();
+
+  /**
+   * Read a coverage from a COG file for the specified geometry.
+   *
+   * @param path Path to the COG file
+   * @param geometry EngineGeometry defining the area of interest
+   * @return GridCoverage2D containing the data within the geometry's bounds
+   * @throws IOException if there is an error reading the file
+   */
+  public static GridCoverage2D getCoverageFromDisk(
+      String path,
+      EngineGeometry geometry
+  ) throws IOException {
+    File file = new File(path);
```

Comment:
If you want an easy way to support both, URL + BufferedReader + InputStreamReader might work. BufferedReader and InputStreamReader work the same way regardless of if the file is remote or local.

Reply:
Actually looks like Geotools takes either a URL or a File - https://github.com/geotools/geotools/blob/main/modules/plugin/geotiff/src/main/java/org/geotools/gce/geotiff/GeoTiffReader.java#L219. You might consider just using URL for local files too: `File.toURI().toURL()`